### DISCUSSION TITLE
integrating/overview

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#13] [General]
# integrating/overview

To integrate Velopack into your application, you must initialise the Velopack as early as possible in app startup, and you should add update checking code somewhere.

https://docs.velopack.io/integrating/overview

**Comment by yprism** [comment]
(unrelated question to this page)
Is it possible to prompt the user to choose a file during installation process? e.g. license file location?

**Reply by caesay** [reply]
Feel free to join our Discord or open an issue on GitHub if you have general questions which are not related to the docs. But to answer your question, it's currently not possible to customise the installer journey or show any custom UI. You can prompt for a license file on your app's first start. There may be more configuration available after https://github.com/velopack/velopack/issues/30 is released, but I'm not sure exactly what that will look like yet.

**Comment by wangdasheng360** [comment]
My asp.net core application needs to be installed into the system service. Can I use Velopack to automatically install it into the system service and update it automatically?

**Reply by caesay** [reply]
Velopack currently (as of July 2024) does not support installing or updating Windows Services.

**Comment by enisn** [comment]
How about roll-back the last update? Or installing a specific version? (The previous version in my case)

**Reply by enisn** [reply]
✅Found the answer in Discord community

Thanks to @caesay

---

> You can list the available releases using `IUpdateSource.GetReleaseFeed`, and provided you want to download/apply one of those, just create a new `UpdateInfo` class given the `VelopackAsset` you'd like to update to, and then follow the usual download/apply steps.

**Comment by verySpicyHeatWave** [comment]
Where would I put the VelopackApp.Build().Run() call in, say, a VB.Net app that doesn't have a "Main" function? Would the main form loading event handler be an appropriate place?

**Reply by caesay** [reply]
It should be as early in your app startup as possible. If you are using WPF or WinForms, you can create a Main function (and not take the default). [There are Velopack samples](https://github.com/velopack/velopack/tree/develop/samples/CSharpWpf) showing how to do this. Alternatively, you could add it to your `Application` constructor or `OnStartup` etc. 

It should not be a part of a form loading handler, as that implies your app startup has already executed a significant amount of code, enough to start showing a window.

**Comment by cyechow** [comment]
I just stumbled across this today while looking at installation frameworks options with good auto-update handling and it looks awesome!

I don't think I saw anything about this in the docs, ClickOnce has a feature where it's able to to launch applications via a URL. This also depends on browser settings/extensions installed but essentially it generates a *.application file which you can upload to a cloud storage location, like S3, and then enter https://example.com/myapp.application which will just launch the application without downloading an installer. It will skip installing the app if it's already there. It's also able to accept query parameters in the URL that gets interpreted when running the application.

Are these things that Velopack can do or will do in the future?

**Reply by caesay** [reply]
It's not possible to replicate this, it's a ClickOnce/Microsoft exclusive as far as I'm aware, so no - not planned.

**Reply by cyechow** [reply]
Just following up on this, the use case for the above is for a web-based application that is intended to integrate/work seamlessly with a desktop application i.e. web app to open the desktop app to continue some user workflow.

In case this is useful for others working with a Windows-based environment, what I did instead is added a custom protocol in the registry under HKCU (user-level). I believe similar things can be done in Linux/MacOS.

**Reply by caesay** [reply]
You can set custom protocols to open web addresses in your application _once it's been installed_. The novelty with ClickOnce is that you can run even if the app is not installed. We can't do what ClickOnce does, but I do plan to add support for url protocols. [See here for more info](https://github.com/velopack/velopack/issues/29).

**Reply by cyechow** [reply]
oh awesome, thanks for the link! yes, I added the custom protocol in as part of `WithFirstRun` and removed it in `WithBeforeUninstallFastCallback`

**Reply by caesay** [reply]
Feel free to share your code or submit a PR if you'd like to help with that issue!

**Comment by Segergren** [comment]
Hey, this might be a dumb question, but is there a way to get the 'current' version number even if there's no update available?

**Reply by Keboo** [reply]
Yes, the `UpdateManager` can show you the currently installed version number. You want to use the CurrentVersion property after checking the IsInstalled property.


Sample here: https://github.com/velopack/velopack/blob/8c2cd109123e01e63e1f1df024e0ddd0dadefb9e/samples/CSharpWpf/MainWindow.xaml.cs#L84

**Comment by BrensNina** [comment]
In fast callbacks I can do something like the code below or  i can't definetly show UI like a message box?:

```cs
[STAThread]
private static void Main(string[] args)
{
    VelopackApp.Build()
        .OnFirstRun(v =>
        {
            // things
        })
        .OnAfterInstallFastCallback(v =>
        {
            string path = @"C:\RandomFolder";

            try
            {
                if (!Directory.Exists(path))
                {
                    Directory.CreateDirectory(path);
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show("Hey Something was wrong, you should know that");
            }
        });
}
```

**Reply by Keboo** [reply]
Though you can, you really should not. The hooks all have [maximum timeouts (15-30 seconds)](https://docs.velopack.io/integrating/hooks). You want to avoid any code that could potentially run logger than that. It is a better idea to simply save the data, and then display the message box on app startup.
